lib/bots/notifybot.ts

```diff
@@ -0,0 +1,918 @@
+		// or the keyframe is incorrect.
+		const repo = this.getRepoDetails(component.repo);
+		const validVersion = (data: string, version: string) => {
+			return data.indexOf(version) !== -1;
+		};
+
+		if (!repo) {
+			throw new Error(`Cannot find appropriate repo/owner for ${component.repo}`);
+		}
+
+		/** Retrieve the changelog. */
+		const fileRequest = {
+			api: this.githubApi,
+			filepath: ChangelogFile,
+			hash: 'HEAD',
+			owner: repo.owner,
+			repo: repo.repo,
+		};
+		return this.retrieveFileFromHash(fileRequest).then((file) => {
+			if (!file) {
+				throw new Error(`Couldn't find the CHANGELOG.md in ${repo.owner}/${repo.repo}`);
+			}
+
+			// We have the Changelog, we want to cull any versions which aren't
+			// relevant by removing them. Do this by splitting the lines based on
+			// the old version and the new version.
+			// For both incrementing and decrementing versions, we want all the versions
+			// that exist between the greatest version (inclusive) and the oldest version
+			// (exclusive).
+			// We can easily get this by swapping the versions for decrementing.
+			let changeParts = [ file ];
+			let changelogString = '';
+
+			// Ensure that the old version (if there is one) and the new version are valid.
+			if (component.oldVersion) {
+				if (!validVersion(file, component.oldVersion)) {
+					this.reportError(new NotifyBotError(NotifyErrorCodes.OldVersionInvalid,
+						`The old version of the component in the old keyframe was invalid`,
+						repo.owner,
+						repo.repo,
+						component.oldVersion
+					));
+					return component;
+				}
+			}
+			if (!validVersion(file, component.newVersion)) {
+				this.reportError(new NotifyBotError(NotifyErrorCodes.NewVersionInvalid,
+					`The new version of the component in the new keyframe was invalid`,
+					repo.owner,
+					repo.repo,
+					component.newVersion
+				));
+
+				return component;
+			}
+
+			// If there's no old version, then *anything* we release is essentially new and
+			// therefore is now implemented. It is implicitly an incrementing version.
+			if (!component.oldVersion) {
+				component.incrementing = true;
+			} else if (component.oldVersion > component.newVersion) {
+				// A new version less than an older version is a version down.
+				component.incrementing = false;
+			} else if (component.newVersion > component.oldVersion) {
+				component.incrementing = true;
+			}
+
+			// If there is an old component, use it to cull anything before that version.
+			if (component.oldVersion) {
+				changeParts = changeParts[0].split(component.incrementing ? component.oldVersion :
+					component.newVersion);
+			}
+			// We are guaranteed a new version, as it's already been verified.
+			changelogString = `## ${component.incrementing ? component.newVersion : component.oldVersion}` +
+				changeParts[0].split(component.incrementing ? component.newVersion : (component.oldVersion || ''))[1];
```

Comment:
It's very hard to tell what this logic is doing, especially with all the interspersed ternary operators and splits. Could we make it clearer? More understandable as something like:

* Split the changelog on new lines
* Find the indexes of the lines matching the new and old versions
* Grab the lines between
* Filter for version headings